### PR TITLE
[onert] Support plugin model loader

### PR DESCRIPTION
**Pull request by hseok-oh**
This commit introduces plugin model loader managing function.

ONE-DCO-1.0-Signed-off-by: Hyeongseok Oh <hseok82.oh@samsung.com>

---

Related issue: https://github.com/Samsung/ONE/issues/12482
Draft: https://github.com/Samsung/ONE/pull/12563

### CODE REVIEW

File: runtime/onert/core/src/loader/ModelLoader.cc [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  // Open custom loader library
+  void *handle = dlopen(libname.c_str(), RTLD_LAZY);
+  if (!handle)
+    throw std::runtime_error("Failed to open " + type + " loader");
+
+  // Get custom loader create function
+  using create_func_t = ILoader *(*)();
+  auto create_fn = reinterpret_cast<create_func_t>(dlsym(handle, "onert_loader_create"));
+  if (!create_fn)
+    throw std::runtime_error("Failed to find loader create function");
```
Comment (YongseopKim) [inline runtime/onert/core/src/loader/ModelLoader.cc:42]: It's okay if there is not `dlclose(handle);`?

Comment (hseok-oh) [reply runtime/onert/core/src/loader/ModelLoader.cc:42]: I'll add `dlclose`

**Review by YongseopKim — DISMISSED** [review]
LGTM

File: runtime/onert/core/src/loader/ModelLoader.cc [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  // Destroy custom loader
+  destroy_fn(loader);
+
+  // Close custom loader library
+  //
+  // NOTE:
+  //  It assumes that custom loader will not be used frequently on runtime session.
+  //  If custom loader is used frequently, it should not close custom loader library and
+  //  save handler to reuse it.
+  dlclose(handle);
```
Comment (ragmani) [inline runtime/onert/core/src/loader/ModelLoader.cc:76]: (Out of curiosity)
Are all objects returned from `dysym()` destroyed when `dlclose()` is called?

Comment (jyoungyun) [reply runtime/onert/core/src/loader/ModelLoader.cc:76]: AFAIK, `dlclose()` decreases the ref count of the dynamic library handle and unloads the library if there is no user of this handle. If there is another user of this handle, the object returned from `dlsym()` can still be accessed. However, the objects returned from `dlsym()` depends on the dynamic library status. Therefore,  we do not need to further destroy the objects returned from `dlsym()`.

**Review by jyoungyun — APPROVED** [review]
LGTM

File: runtime/onert/core/include/loader/ModelLoader.h [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef __ONERT_LOADER_LOADERS_H__
+#define __ONERT_LOADER_LOADERS_H__
```
Comment (glistening) [inline runtime/onert/core/include/loader/ModelLoader.h:18]: @hseok-oh The guard seems not follow our convention. Is there specific reason what I miss?

**Review by ragmani — APPROVED** [review]
LGTM

File: runtime/onert/core/include/loader/ModelLoader.h (hunk 3 above)
Comment (hseok-oh) [reply runtime/onert/core/include/loader/ModelLoader.h:18]: No. It's mistake. I'll update.

File: runtime/onert/core/include/loader/ModelLoader.h [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+namespace onert
+{
+namespace loader
+{
+
+/**
+ * @brief     Create custom loader and load model from file
+ * @param[in] filename  File path to load model from
+ * @param[in] type      Type of custom loader to create
```
Comment (glistening) [inline runtime/onert/core/include/loader/ModelLoader.h:32]: (optional) This method is `loadModel`, and the type of custom loader seems implementation-detail, which is better to hide. I prefer `Type of model to load`.
```suggestion
 * @param[in] type      Type of model to load
```

**Review by glistening — DISMISSED** [review]
LGTM

**Review by YongseopKim — APPROVED** [review]
LGTM 👍

**Review by ragmani — DISMISSED** [review]

**Review by ragmani — APPROVED** [review]

**Review by YongseopKim — APPROVED** [review]

**Review by glistening — APPROVED** [review]
LGTM

**Review by jyoungyun — APPROVED** [review]
LGTM

**Pull request merged**